Update FNet Fourier Transform

gchhablani opened the pull request:
# What does this PR do?

This PR changes the Fourier Transform according to the [latest updates](https://github.com/google-research/google-research/commit/4f7815fe3aaf71b811a0fb82ac8768a4e8ab0428#diff-4a2be83f4d2b702b206d3f64b2aeb209ac7a2d65fd28382916225463b984bc2c) to the FNet repository.

I will update the configurations of the checkpoints if the changes look fine.

@patrickvonplaten @LysandreJik @sgugger @patil-suraj

gchhablani committed:
Update FNet Fourier Transform

sgugger commented on src/transformers/models/fnet/configuration_fnet.py:
```diff
@@ -64,13 +64,13 @@ class FNetConfig(PretrainedConfig):
-            The sequence length that is expected by the model when using TPUs. This will be used to initialize the DFT
-            matrix only when `use_tpu_fourier_optimizations` is set to :obj:`True` and the input sequence is shorter
-            than or equal to 4096 tokens.
+        use_fft (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Determines whether to use n-dimensional FFTs. If :obj:`True`, the model will use n-dimensional FFT
+            transforms. Set to :obj:`False` for using DFT matrix multiplication for fourier transform.
+        actual_seq_length (:obj:`int`, `optional`, defaults to 512):
+            The sequence length that is expected by the model. This will be used to initialize the DFT matrix when
+            `use_fft` is set to :obj:`False`. Additionally, this is used for validation when using FFT. The recommended
+            values for lengths greater than 4096 is powers of 2.
```
Those are breaking changes and cannot be accepted.

patrickvonplaten replied:
@gchhablani - could you leave the old arguments and simply add new ones while making sure that when the old arguments are used with the default new arguments the behavior stays 1-to-1 the same. This would be fine!

sgugger reviewed (commented):
Thanks for your PR, but we can't do breaking changes like this in the Transformers codebase. While researchers can certainly update their code as they want, the models of the library are used in production and our users trust us to strictly maintain backward compatibility.

patrickvonplaten commented:
Thanks for the improved version of the FNet Fourier transform! In order to keep backwards compatibility, we should try to leave all existing arguments and simply add new ones - this would definitely be possible :-)

gchhablani commented:
@sgugger @patrickvonplaten
Does this mean I cannot remove the custom `fftn` function?

Or just the configuration arguments?
If I keep the old config arguments, won't I have to keep the relevant code as well?

sgugger commented:
No, the old function can't be removed for backward compatibility, sadly. We can only add new arguments that trigger some new behavior.

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into update_fnet

gchhablani commented:
@sgugger I have updated the code accordingly. Added another config parameter `use_latest` to allow switching. By default it is `False`. Does this look fine?

Also, is it possible to deprecate the previous version later and throw a similar warning, then?

gchhablani committed:
Add changes from review

gchhablani committed:
Fix logging warning bug

sgugger commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -166,26 +166,54 @@ def __init__(self, config):
+                    raise ValueError(
+                        f"For larger seuence lengths (>4096), the actual sequence length {self.config.actual_seq_length} must be a power of 2."
+                    )
+                self.fourier_transform = partial(torch.fft.fftn, dim=(1, 2))
```
How recent is this? Should there be a PyTorch version check?

gchhablani replied:
@sgugger Do you mean `torch.fft.fftn`?

gchhablani replied:
@sgugger The fftn function [was added in PyTorch 1.7](https://github.com/pytorch/pytorch/releases/tag/v1.7.0), in this [PR](https://github.com/pytorch/pytorch/pull/44550).

patrickvonplaten replied:
Think a version check would indeed be very nice here. @gchhablani - could you add it here: https://github.com/huggingface/transformers/blob/41436d3dfb98e0d17f018db29790b65663358edf/src/transformers/optimization.py#L301

sgugger reviewed (approved):
Thanks a lot for adapting your code and making it backward compatible! I just have one last comment on my side.

gchhablani committed:
Update use_latest docs

gchhablani commented:
@sgugger I will fix the test according to the new change.

Also, wanted to ask if there is a possiblity of ever deprecating the old version? Just out of curiosity.

sgugger commented:
For modeling code, it's tougher to go through the usual deprecation cycle as we might have models on the Hub using the old code. They would then stop working if we ever removed the old code, which is something we try very hard to avoid.

gchhablani commented:
@sgugger Also, what do you think about [this comment](https://github.com/huggingface/transformers/issues/13684#issuecomment-924472556)?

CC @patrickvonplaten

gchhablani committed:
Fix modeling file for backward compatibility

stefan-it commented:
@gchhablani the `model.half` support would be cool to have, I was missing it in my last FNet GPU training 😅

sgugger commented:
Can you put the fix for `model.half()` in a new PR @gchhablani ? For the other part of making the attributes parameters, I'm not opposed to do it as long as it's backward compatible.

gchhablani commented:
@sgugger I checked by doing `model.cuda()` on the current model. Refer to [this colab notebook](https://colab.research.google.com/drive/1fB5GOGuH3xClyTXJ_zagR_G-3Qd7QzMQ?usp=sharing) if needed. The buffers are also pushed to the `cuda:0` device. I don't think any change is needed regarding this. Wdyt?

gchhablani commented:
@stefan-it Can you please explain the issue you face when you do `model.half()`? In this notebook linked in the previous comment, I have tried `model.half()` as well. Except the buffers, everything becomes `torch.float16`.

sgugger commented:
If the buffers are properly pushed, then there is no need to change anything I believe.

patrickvonplaten commented:
Looks good to me as well! @gchhablani - do you think the new version yields a significant speed-up? In this case it might be worth adding a deprecation warning to the current config.

gchhablani commented:
@patrickvonplaten With the current changes, there is almost negligible modification with `use_latest`. With default conditions (for GPU), we would be using the `torch.fft.fftn` method as before. 

Not sure how much difference is there between the custom fftn (fft in loop) and the torch method, speed-wise. 

I just thought that we might need to update to the latest changes as on the original repository.

gchhablani commented:
CC @ontocord

ontocord commented:
Ha! Great to see you all are adding some of this stuff...

stefan-it commented:
Hi @gchhablani , some feedback:

```bash
Traceback (most recent call last):
  File "run_mlm.py", line 552, in <module>
    main()
  File "run_mlm.py", line 501, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/trainer.py", line 1312, in train
    tr_loss_step = self.training_step(model, inputs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/trainer.py", line 1839, in training_step
    loss = self.compute_loss(model, inputs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/trainer.py", line 1873, in compute_loss
    outputs = model(**inputs)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 792, in forward
    outputs = self.fnet(
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 636, in forward
    encoder_outputs = self.encoder(
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 334, in forward
    layer_outputs = layer_module(hidden_states)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 293, in forward
    self_fourier_outputs = self.fourier(hidden_states)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 246, in forward
    self_outputs = self.self(hidden_states)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/turkish-fnet/transformers-fnet-refactoring/src/transformers/models/fnet/modeling_fnet.py", line 225, in forward
    outputs = self.fourier_transform(hidden_states).real
RuntimeError: Unsupported dtype Half
```
Occurs when using the (PyTorch) language modeling example with `--fp16` option - even with the latest `update_fnet` branch.

gchhablani commented:
@ontocord Can you please explain why you think that the DFT matrices need to made into parameters?

@patrickvonplaten @sgugger Any idea on how can I fix the issue @stefan-it just mentioned?

ontocord commented:
They do not convert to cuda or half automatically unless they are in the same parameter list of a module. I made them no_grad Params. 

> On Oct 1, 2021, at 2:26 PM, Gunjan Chhablani ***@***.***> wrote:
> 
> ﻿
> @ontocord Can you please explain why you think that the DFT matrices need to made into parameters?
> 
> @patrickvonplaten @sgugger Any idea on how can I fix the issue @stefan-it just mentioned?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

gchhablani commented:
Unstale
